cmd/ingest/ingest.go

```diff
@@ -42,24 +57,162 @@ func main() {
 	var message Message
 
 	go func() {
-		for d := range broker.GetMessages(mq, config.Broker.Queue) {
-			log.Debugf("received a message: %s", d.Body)
-			err := json.Unmarshal(bytes.Replace(d.Body, []byte(`'`), []byte(`"`), -1), &message)
-			if err != nil {
-				log.Errorf("Not a json message: %s", err)
+		for delivered := range broker.GetMessages(mq, config.Broker.Queue) {
+			log.Debugf("Received a message: %s", delivered.Body)
+			if e := json.Unmarshal(delivered.Body, &message); e != nil {
+				log.Errorf("Not a json message: %s", e)
 			}
 
-			e := reflect.ValueOf(&message).Elem()
-			for i := 0; i < e.NumField(); i++ {
-				if e.Field(i).Interface() == "" || e.Field(i).Interface() == 0 {
-					log.Errorf("%s is missing", e.Type().Field(i).Name)
-					err = fmt.Errorf("%s is missing", e.Type().Field(i).Name)
+			fileID, ife := db.InsertFile(fmt.Sprintf("%s/%s", config.Inbox.Location, message.FilePath), message.User)
+			if ife != nil {
+				log.Errorf("InsertFile failed, reason: %v", ife)
+				// This should really be hadled by the DB retry mechanism
+			}
+
+			file, rfe := storage.ReadFile(config.Inbox.Type, fmt.Sprintf("%s/%s", config.Inbox.Location, message.FilePath))
+			if rfe != nil {
+				log.Errorf("Failed to open file: %s, reason: %v", fmt.Sprintf("%s/%s", config.Inbox.Location, message.FilePath), rfe)
+				continue
+			}
+
+			fileSize, gfse := storage.GetFileSize(config.Inbox.Type, fmt.Sprintf("%s/%s", config.Inbox.Location, message.FilePath))
+			if gfse != nil {
+				log.Errorf("Failed to get file size: %s, reason: %v", fmt.Sprintf("%s/%s", config.Archive.Location, message.FilePath), gfse)
+				continue
+			}
+
+			hash := sha256.New()
+			// 512KiB readbuffer, this must be large enough that we get the entire header and the first 64KiB datablock 
+			readBuffer := make([]byte, 512 * 1024)
```

Comment:
After a day of consideration, I'm still a bit skeptical about the fixed-sized buffer approach. Files can be very large. For example, our test submission is 40+GB, which means that encrypted it would contain ~ 625000 data blocks. And according to Crypt4GH specification, each data block can be theoretically encrypted with a separate header packet, which would lead to ~ 625000 header packets in this case. I know, it would be a very special case, but I think that we need to respect Crypt4GH standard anyway. And this header will not fit into the remaining 512 - 64 = 448 KBs (448000 / 625000 = less than a byte per header). So, IMHO, this approach is both not very neat in terms of code, but also simply not 100% reliable. I'd argue for opening the stream twice (as was suggested in Slack).

Reply:
But it can work for now, as a minimally viable solution. And can be reworked later.

Reply:
In the end it will be configurable in order to match the S3 chunk size and then we are talking ~8-32MiB buffer size.